### PR TITLE
better error for gradient check

### PR DESCRIPTION
**Pull request by DeNeutoy**
The assert gave a really rubbish error in the case that the gradient was `None`, so this will improve that.

### CODE REVIEW

File: allennlp/common/testing/model_test_case.py [hunk 1]
```diff
@@ -133,16 +133,25 @@ def check_model_computes_gradients_correctly(model, model_batch):
...
+        for name, parameter in model.named_parameters():
             zeros = torch.zeros(parameter.size())
             if parameter.requires_grad:
+
+                if parameter.grad is None:
+                    has_zero_or_none_grads[name] = "No gradient computed (i.e parameter.grad is None)"
                 # Some parameters will only be partially updated,
                 # like embeddings, so we just check that any gradient is non-zero.
-                assert (parameter.grad.data.cpu() != zeros).any()
+                if not (parameter.grad.data.cpu() != zeros).any():
```
Comment (joelgrus) [inline allennlp/common/testing/model_test_case.py]: is there a way to rewrite this? `not (grad != zeros).any()` is really hard to understand

Comment (joelgrus) [reply allennlp/common/testing/model_test_case.py]: like, would

```
if (parameter.grad.data.cpu() == 0).all()
```

work?

File: allennlp/common/testing/model_test_case.py [hunk 2]
```diff
@@ -133,16 +133,25 @@ def check_model_computes_gradients_correctly(model, model_batch):
...
-                assert (parameter.grad.data.cpu() != zeros).any()
+                if (parameter.grad.data.cpu() == zeros).all():
+                    has_zero_or_none_grads[name] = f"zeros with shape ({tuple(parameter.grad.size())})"
             else:
                 assert parameter.grad is None
 
+        if has_zero_or_none_grads:
+            for name, grad in has_zero_or_none_grads.items():
+                print(f"Parameter: {name} had incorrect gradient: {grad}")
+            raise Exception("Incorrect gradients found. See stack trace for info.")
```
Comment (matt-gardner) [inline allennlp/common/testing/model_test_case.py]: The info's in stdout, not the stack trace, right?
